### PR TITLE
[SPARK-36704][CORE] Expand exception handling to more Java 9 cases where reflection is limited at runtime, when reflecting to manage DirectByteBuffer settings

### PR DESCRIPTION
**Pull request by srowen**
### What changes were proposed in this pull request?

Improve exception handling in the Platform initialization, where it attempts to assess whether reflection is possible to modify DirectByteBuffer. This can apparently fail in more cases on Java 9+ than are currently handled, whereas Spark can continue without reflection if needed.

More detailed comments on the change inline.

### Why are the changes needed?

This exception seems to be possible and fails startup:

```
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make private java.nio.DirectByteBuffer(long,int) accessible: module java.base does not "opens java.nio" to unnamed module @71e9ddb4
        at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:357)
        at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
        at java.base/java.lang.reflect.Constructor.checkCanSetAccessible(Constructor.java:188)
        at java.base/java.lang.reflect.Constructor.setAccessible(Constructor.java:181)
        at org.apache.spark.unsafe.Platform.<clinit>(Platform.java:56)
```

### Does this PR introduce _any_ user-facing change?

Should strictly allow Spark to continue in more cases.

### How was this patch tested?

Existing tests.

### CODE REVIEW

File: common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java [hunk 1]
```diff
@@ -77,28 +67,52 @@
...
     try {
-      Class<?> cleanerClass = Class.forName(cleanerClassName);
-      Method createMethod = cleanerClass.getMethod("create", Object.class, Runnable.class);
-      // Accessing jdk.internal.ref.Cleaner should actually fail by default in JDK 9+,
-      // unfortunately, unless the user has allowed access with something like
-      // --add-opens java.base/java.lang=ALL-UNNAMED  If not, we can't really use the Cleaner
-      // hack below. It doesn't break, just means the user might run into the default JVM limit
-      // on off-heap memory and increase it or set the flag above. This tests whether it's
-      // available:
+      Class<?> cls = Class.forName("java.nio.DirectByteBuffer");
```
Comment (srowen) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:70]: This was simply moved from above to further unify handling of exceptions during init

File: common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java [hunk 2]
```diff
@@ -77,28 +67,52 @@
...
-        // Don't throw an exception, but can't log here?
-        createMethod = null;
-      } catch (InvocationTargetException ite) {
-        // shouldn't happen; report it
-        throw new IllegalStateException(ite);
+        constructor.setAccessible(true);
+        cleanerField.setAccessible(true);
+      } catch (RuntimeException re) {
+        // This is a Java 9+ exception, so needs to be handled without importing it
+        if ("InaccessibleObjectException".equals(re.getClass().getSimpleName())) {
```
Comment (srowen) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:78]: This is really the new exception handling; now have to catch this and also handle the case where DBB_CONSTRUCTOR and DBB_CLEANER_FIELD aren't available.

**Comment by srowen** [comment]
If this is OK, this should go into 3.2.x for 3.2.0 if possible

**Comment by dongjoon-hyun** [comment]
Thank you, @srowen . +1 for backporting from my side. cc @gengliangwang 
> If this is OK, this should go into 3.2.x for 3.2.0 if possible

File: common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java [hunk 3]
```diff
@@ -77,28 +67,52 @@
...
+          // Continue, but the constructor/field are not available
+          // See comment below for more context
+          constructor = null;
+          cleanerField = null;
```
Comment (mridulm) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:82]: Given the `throw`, why do we need this ? Class wont be initialized right ?

**Comment by mridulm** [comment]
Rest of the changes look fine, +1 to backporting to 3.2

File: common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java (hunk 3 above)
Comment (srowen) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:82]: This block helps ensure the class does initialize even if reflection can't be used on DirectByteBuffer. Nulling these particular values out doesn't matter a whole lot; without it, the two static final variables do get set to values that can't be used anyway. It seemed tidier and more symmetric to just ensure they init to null in this case, though later they won't even be used if the rest didn't succeed in setting up values necessary to make reflection work.

**Comment by gengliangwang** [comment]
+1 to backporting to 3.2, too

File: common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java (hunk 3 above)
Comment (mridulm) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:82]: I am missing something here, given [this](https://github.com/apache/spark/blob/09d72879a9752976f11fbce6ed7be224ecb8fba0/common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java#L84), `DBB_CONSTRUCTOR` and `DBB_CLEANER_FIELD` are not set.
And as the exception is thrown from the static block, the class `Platform` will fail to load.

To illustrate:
```
$ cat Test.java 

public class Test {
  public static void main(String[] args) {
    System.out.println("Value = " + TestStatic.value);
  }
}

class TestStatic {
  static final String value;

  static {
    if (Boolean.getBoolean("THROW_EXCEPTION")) {
      throw new RuntimeException("failed");
    }
    value = "hi";
  }
}

$ java -cp . -DTHROW_EXCEPTION=true Test
Exception in thread "main" java.lang.ExceptionInInitializerError
	at Test.main(Test.java:4)
Caused by: java.lang.RuntimeException: failed
	at TestStatic.<clinit>(Test.java:13)
	... 1 more

```

Comment (srowen) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:82]: Ugh, yeah I made a basic error here - needs to be 'else throw re'. 
I should also say I don't know how this arises, so can't test for it. It came up on the list and seems therefore possible, but also didn't come up before in testing

**Review by mridulm — APPROVED** [review]
LGTM, thanks for fixing this @srowen !

**Comment by srowen** [comment]
Merged to master/3.2/3.1/3.0

**Pull request closed without merging**